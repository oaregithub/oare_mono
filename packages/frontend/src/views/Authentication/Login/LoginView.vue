<template>
  <OareCard title="Sign In">
    <v-text-field
      v-model="email"
      class="test-email"
      :label="$t('login.email')"
      outlined
    />
    <v-text-field
      class="test-password"
      outlined
      @keyup.enter="logIn"
      v-model="password"
      :label="$t('login.password')"
      type="password"
    />
    <p class="subtitle error--text test-error-text">{{ errorMsg }}</p>
    <v-btn text class="text-none" to="/send_reset_password_email"
      >Forgot your password?</v-btn
    >
    <v-btn text class="text-none" to="/register">{{
      $t('login.dontHaveAccount')
    }}</v-btn>
    <br />

    <template #actions>
      <OareLoaderButton
        class="text-right test-signin-btn"
        @click="logIn"
        color="primary"
        :loading="loadings.signInButton"
        :disabled="!email || !password"
      >
        {{ $t('login.signIn') }}
      </OareLoaderButton>
    </template>
  </OareCard>
</template>

<script lang="ts">
import { defineComponent, ref, PropType } from '@vue/composition-api';
import { NavigationGuard } from 'vue-router';
import sl from '@/serviceLocator';
import defaultRouter from '../../../router';
import Router from 'vue-router';

export interface FormRef {
  validate: () => void;
}

const beforeRouteEnter: NavigationGuard = (to, from, next) => {
  next(() => {
    const store = sl.get('store');
    if (store.getters.isAuthenticated) {
      defaultRouter.push('/');
    }
  });
};

export default defineComponent({
  name: 'LoginView',
  props: {
    router: {
      type: Object as PropType<Router>,
      default: () => defaultRouter,
    },
  },
  beforeRouteEnter,
  setup({ router }) {
    const store = sl.get('store');
    const server = sl.get('serverProxy');

    const email = ref('');
    const password = ref('');
    const errorMsg = ref('');
    const loadings = ref({ signInButton: false });

    const logIn = async () => {
      loadings.value.signInButton = true;
      errorMsg.value = '';
      let userData = {
        email: email.value,
        password: password.value,
      };

      try {
        const response = await server.login(userData);
        store.setUser(response);
        const permissions = await server.getUserPermissions();
        store.setPermissions(permissions);
        router.push('/');
      } catch (err) {
        errorMsg.value = err.response.data.message;
      } finally {
        loadings.value.signInButton = false;
      }
    };

    return {
      email,
      password,
      errorMsg,
      loadings,
      logIn,
    };
  },
});
</script>
