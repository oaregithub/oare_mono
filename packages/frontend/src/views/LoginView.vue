<template>
  <v-card class="mt-12 pa-3 mx-auto" max-width="600" elevation="24">
    <v-card-title>
      <h3 class="oare-header mb-2">{{ $t("login.signIn") }}</h3>
    </v-card-title>
    <v-card-text>
      <v-form v-model="valid" ref="form">
        <v-text-field
          v-model="email"
          class="emailField"
          :rules="emailRules"
          :label="$t('login.email')"
          outlined
          data-login-email
        />
        <v-text-field
          class="passwordField"
          outlined
          @keyup.enter="logIn"
          v-model="password"
          :rules="passwordRules"
          :label="$t('login.password')"
          type="password"
          data-login-password
        />
      </v-form>
      <p class="subtitle error--text">{{ errorMsg }}</p>
      <v-btn text class="text-none" to="/register">{{
        $t("login.dontHaveAccount")
      }}</v-btn>
      <br />
    </v-card-text>

    <v-card-actions>
      <OareLoaderButton
        class="text-right loginBtn"
        @click="logIn"
        color="primary"
        :loading="loadings.signInButton"
      >
        {{ $t("login.signIn") }}
      </OareLoaderButton>
    </v-card-actions>
  </v-card>
</template>

<script lang="ts">
import { defineComponent, ref, Ref } from "@vue/composition-api";
import { NavigationGuard } from "vue-router";
import i18n from "../i18n";
import store from "../store";
import router from "../router";

export interface FormRef {
  validate: () => void;
}

const beforeRouteEnter: NavigationGuard = (to, from, next) => {
  next(() => {
    if (store.getters.isAuthenticated) {
      router.push("/");
    }
  });
};

export default defineComponent({
  name: "LoginView",

  beforeRouteEnter,
  setup() {
    const valid = ref(false);
    const email = ref("");
    const password = ref("");
    const errorMsg = ref("");
    const emailRules = ref([(v: string) => !!v || i18n.t("login.emailError")]);
    const passwordRules = ref([
      (v: string) => !!v || i18n.t("login.passwordError"),
    ]);
    const loadings = ref({ signInButton: false });
    const form: Ref<null | FormRef> = ref(null);

    const logIn = async () => {
      loadings.value.signInButton = true;
      if (form.value) {
        form.value.validate();
      }
      if (!valid.value) {
        errorMsg.value = "Please fill in all required fields.";
        loadings.value.signInButton = false;
        return;
      }
      errorMsg.value = "";
      let userData = {
        email: email.value,
        password: password.value,
      };

      try {
        await store.dispatch("login", userData);
        router.push("/");
      } catch (err) {
        errorMsg.value = err;
      }
      loadings.value.signInButton = false;
    };

    return {
      valid,
      email,
      password,
      errorMsg,
      emailRules,
      passwordRules,
      loadings,
      form,
      logIn,
    };
  },
});
</script>
