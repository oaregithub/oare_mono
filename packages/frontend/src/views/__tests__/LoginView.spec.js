import Vuetify from "vuetify";
import VueCompositionApi from "@vue/composition-api";
import { mount, createLocalVue } from "@vue/test-utils";
import LoginView from "../LoginView.vue";
import flushPromises from "flush-promises";

const vuetify = new Vuetify();
const localVue = createLocalVue();
localVue.use(VueCompositionApi);

describe("LoginView test", () => {
  const mockProps = {
    router: {
      push: jest.fn(),
    },
    store: {
      dispatch: jest.fn().mockResolvedValue(true),
    },
  };
  const createWrapper = (props = mockProps) =>
    mount(LoginView, {
      localVue,
      vuetify,
      mocks: {
        $t: jest.fn(),
      },
      propsData: props,
      stubs: ["router-link"],
    });

  it("matches snapshot", () => {
    expect(createWrapper()).toMatchSnapshot();
  });

  it("redirects to home page on successful login", async () => {
    const wrapper = createWrapper();
    const emailInput = wrapper.find(".test-email input");
    const passwordInput = wrapper.find(".test-password input");

    await emailInput.setValue("myemail@test.com");
    await passwordInput.setValue("password");

    const signinBtn = wrapper.find(".test-signin-btn");
    await signinBtn.trigger("click");
    await flushPromises();

    expect(mockProps.store.dispatch).toHaveBeenCalledWith("login", {
      email: "myemail@test.com",
      password: "password",
    });
  });

  it("sign in button is only active when fields are filled in", async () => {
    const wrapper = createWrapper();

    const signinButton = wrapper.find(".test-signin-btn");
    expect(signinButton.element).toBeDisabled();

    const emailInput = wrapper.find(".test-email input");
    await emailInput.setValue("email");
    expect(signinButton.element).toBeDisabled();

    const passwordInput = wrapper.find(".test-password input");
    await passwordInput.setValue("password");
    expect(signinButton.element).toBeEnabled();
  });

  it("displays error message on unsuccessful login", async () => {
    const wrapper = createWrapper({
      ...mockProps,
      store: {
        dispatch: jest.fn().mockRejectedValue("Invalid login"),
      },
    });

    const emailInput = wrapper.find(".test-email input");
    const passwordInput = wrapper.find(".test-password input");
    const signinButton = wrapper.find(".test-signin-btn");

    await emailInput.setValue("email");
    await passwordInput.setValue("password");
    await signinButton.trigger("click");
    await flushPromises();

    const errorText = wrapper.find(".test-error-text");
    expect(errorText.text()).toBe("Invalid login");
  });
});
