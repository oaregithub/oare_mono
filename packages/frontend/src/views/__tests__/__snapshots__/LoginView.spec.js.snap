// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginView test matches snapshot 1`] = `
<div class="mt-12 pa-3 mx-auto v-card v-sheet theme--light elevation-24" style="max-width: 600px;">
  <div class="v-card__title">
    <h3 class="oare-header mb-2">Sign In</h3>
  </div>
  <div class="v-card__text d-flex flex-column">
    <div class="v-input test-email theme--light v-text-field v-text-field--enclosed v-text-field--outlined">
      <div class="v-input__control">
        <div class="v-input__slot">
          <fieldset aria-hidden="true">
            <legend style="width: 0px;"><span>​</span></legend>
          </fieldset>
          <div class="v-text-field__slot"><input id="input-5" type="text"></div>
        </div>
        <div class="v-text-field__details">
          <div class="v-messages theme--light">
            <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
          </div>
        </div>
      </div>
    </div>
    <div class="v-input test-password theme--light v-text-field v-text-field--enclosed v-text-field--outlined">
      <div class="v-input__control">
        <div class="v-input__slot">
          <fieldset aria-hidden="true">
            <legend style="width: 0px;"><span>​</span></legend>
          </fieldset>
          <div class="v-text-field__slot"><input id="input-8" type="password"></div>
        </div>
        <div class="v-text-field__details">
          <div class="v-messages theme--light">
            <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
          </div>
        </div>
      </div>
    </div>
    <p class="subtitle error--text test-error-text"></p>
    <router-link-stub to="/send_reset_password_email" tag="a" activeclass="v-btn--active" exactactiveclass="v-btn--active" ariacurrentvalue="page" event="click" class="text-none v-btn v-btn--flat v-btn--router v-btn--text theme--light v-size--default"><span class="v-btn__content">Forgot your password?</span></router-link-stub>
    <router-link-stub to="/register" tag="a" activeclass="v-btn--active" exactactiveclass="v-btn--active" ariacurrentvalue="page" event="click" class="text-none v-btn v-btn--flat v-btn--router v-btn--text theme--light v-size--default"><span class="v-btn__content"></span></router-link-stub> <br>
  </div>
  <div class="v-card__actions"><button type="button" disabled="disabled" class="text-right test-signin-btn v-btn v-btn--contained v-btn--disabled theme--light v-size--default" color="primary"><span class="v-btn__content">

    </span></button></div>
</div>
`;
